jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.cpp

```diff
@@ -743,7 +744,634 @@ ecma_builtin_string_prototype_object_match (ecma_value_t this_arg, /**< this arg
+          lit_utf8_iterator_t nested_search_iterator = search_iterator;
+          lit_utf8_iterator_t nested_input_iterator = input_iterator;
+          match_end = match_start + 1;
+
+          match_found = true;
+          while (!lit_utf8_iterator_is_eos (&nested_search_iterator))
+          {
+            if (lit_utf8_iterator_is_eos (&nested_input_iterator))
+            {
+              match_found = false;
+              break;
+            }
+
+            ecma_char_t search_character = lit_utf8_iterator_read_next (&nested_search_iterator);
+            ecma_char_t input_character = lit_utf8_iterator_read_next (&nested_input_iterator);
+
+            if (search_character != input_character)
+            {
+              match_found = false;
+              break;
+            }
+            match_end++;
+          }
+
+          if (match_found)
+          {
+            break;
+          }
+        }
+        match_start++;
+      }
+    }
+
+    if (match_found)
+    {
+      ecma_value_t arguments_list_p[1] = { context_p->regexp_or_search_string };
+      ECMA_TRY_CATCH (new_array_value,
+                      ecma_op_create_array_object (arguments_list_p, 1, false),
+                      ret_value);
+
+      context_p->match_start = match_start;
+      context_p->match_end = match_end;
+
+      ret_value = ecma_make_normal_completion_value (ecma_copy_value (new_array_value, true));
+
+      ECMA_FINALIZE (new_array_value);
+    }
+    else
+    {
+      ret_value = ecma_make_simple_completion_value (ECMA_SIMPLE_VALUE_NULL);
+    }
+
+    MEM_FINALIZE_LOCAL_ARRAY (input_start_p);
+    MEM_FINALIZE_LOCAL_ARRAY (search_start_p);
+  }
+  return ret_value;
+} /* ecma_builtin_string_prototype_object_replace_match */
+
+/**
+ * Generic helper function to construct the string which replaces the matched part
+ *
+ * @return completion value
+ *         Returned value must be freed with ecma_free_completion_value.
+ */
+static ecma_completion_value_t
+ecma_builtin_string_prototype_object_replace_get_string (replace_search_context *context_p, /**< search context */
+                                                         ecma_value_t match_value) /**< returned match value */
+{
+  ecma_completion_value_t ret_value = ecma_make_empty_completion_value ();
+
+  if (context_p->is_replace_callable)
+  {
+    ecma_object_t *match_object_p = ecma_get_object_from_value (match_value);
+    ecma_string_t *length_string_p = ecma_get_magic_string (LIT_MAGIC_STRING_LENGTH);
+
+    ECMA_TRY_CATCH (length_value,
+                    ecma_op_object_get (match_object_p, length_string_p),
+                    ret_value);
+
+    JERRY_ASSERT (ecma_is_value_number (length_value));
+
+    ecma_length_t length = (ecma_length_t) (*ecma_get_number_from_value (length_value));
+
+    JERRY_ASSERT (length >= 1);
+
+    MEM_DEFINE_LOCAL_ARRAY (arguments_list,
+                            length + 2,
```

Comment:
I am not sure I follow you here as well. You mean length is maximum length - 1, and length + 2 overflows? This problem is more general, e.g. if the allocated length is less than maximum, but the sizeof allocated type is not 1, it can also be an overflow. I mean if we allocate max_length - 1 ecma_value-s, the (max_length-1)*sizeof(ecma_value) also overflows. So this is definitely needs to be solved in the macro if we want it.

Am I understand correctly that you plan to change ecma_length to uint16?